cpp/include/raft/spectral/kmeans.hpp

```diff
@@ -0,0 +1,898 @@
+  gridDim.y = 1;
+  gridDim.z = 1;
+  minDistances<<<gridDim, blockDim, 0, stream>>>(n, k, dists, codes,
+                                                 clusterSizes);
+  CHECK_CUDA(stream);
+
+  // Compute residual sum of squares
+  *residual_host =
+    thrust::reduce(thrust_exec_policy, thrust::device_pointer_cast(dists),
+                   thrust::device_pointer_cast(dists + n));
+
+  return 0;
+}
+
+/// Update cluster centroids for k-means algorithm
+/** All clusters are assumed to be non-empty.
+ *
+ *  @param n Number of observation vectors.
+ *  @param d Dimension of observation vectors.
+ *  @param k Number of clusters.
+ *  @param obs (Input, device memory, d*n entries) Observation
+ *    matrix. Matrix is stored column-major and each column is an
+ *    observation vector. Matrix dimensions are d x n.
+ *  @param codes (Input, device memory, n entries) Cluster
+ *    assignments.
+ *  @param clusterSizes (Input, device memory, k entries) Number of
+ *    points in each cluster.
+ *  @param centroids (Output, device memory, d*k entries) Centroid
+ *    matrix. Matrix is stored column-major and each column is a
+ *    centroid. Matrix dimensions are d x k.
+ *  @param work (Output, device memory, n*d entries) Workspace.
+ *  @param work_int (Output, device memory, 2*d*n entries)
+ *    Workspace.
+ *  @return Zero if successful. Otherwise non-zero.
+ */
+template <typename IndexType_, typename ValueType_, typename ThrustExePolicy>
+static int updateCentroids(handle_t const& handle,
+                           ThrustExePolicy thrust_exec_policy, IndexType_ n,
+                           IndexType_ d, IndexType_ k,
+                           const ValueType_* __restrict__ obs,
+                           const IndexType_* __restrict__ codes,
+                           const IndexType_* __restrict__ clusterSizes,
+                           ValueType_* __restrict__ centroids,
+                           ValueType_* __restrict__ work,
+                           IndexType_* __restrict__ work_int) {
+  // -------------------------------------------------------
+  // Variable declarations
+  // -------------------------------------------------------
+
+  // Useful constants
+  const ValueType_ one = 1;
+  const ValueType_ zero = 0;
+
+  auto cublas_h = handle.get_cublas_handle();
+  auto stream = handle.get_stream();
+
+  // CUDA grid dimensions
+  dim3 blockDim, gridDim;
+
+  // Device memory
+  thrust::device_ptr<ValueType_> obs_copy(work);
+  thrust::device_ptr<IndexType_> codes_copy(work_int);
+  thrust::device_ptr<IndexType_> rows(work_int + d * n);
+
+  // Take transpose of observation matrix
+  CUBLAS_CHECK(cublasgeam(cublas_h, CUBLAS_OP_T, CUBLAS_OP_N, n, d, &one, obs,
+                          d, &zero, (ValueType_*)NULL, n,
+                          thrust::raw_pointer_cast(obs_copy), n, stream));
+
+  // Cluster assigned to each observation matrix entry
+  thrust::sequence(thrust_exec_policy, rows, rows + d * n);
```

Comment:
We cannot make that assumption. Again, that constraint can only be imposed by the caller. The caller should document that at call point.

Reply:
The main reason I'm saying we should at least document the requirement / implication here is because of the thrust exec policy and the stream on the handle are not the same then this algorithm could break in very unexpected ways. Parts like the above are not expecting to run concurrently, so for example if we have a cudaMemcpyAsync in one stream and a bunch of thrust operations in another then this could give unexpected behaviors. Best to document this the requirement here so the user knows this implication.

Reply:
Okay, done.